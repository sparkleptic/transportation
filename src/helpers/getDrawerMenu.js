// @flow

import convertArrayToMap from './convertArrayToMap';

import type {DrawerMenu} from '../layout/Header/DrawerList';

type Permission = {
  moduleCode: string,
  moduleName: string,
  orderNo: string,
};

export default function getDrawerMenu(
  drawerMenuList: Array<DrawerMenu>,
  permissionList: Array<Permission>,
) {
  let mappedDrawerMenuList = convertArrayToMap(drawerMenuList, 'moduleCode');
  let sortedPermissionList = permissionList.sort((a, b) => {
    let {orderNo: orderNoA} = a;
    let {orderNo: orderNoB} = b;
    if (orderNoA && orderNoB && orderNoA.length < orderNoB.length) {
      return -1;
    } else if (orderNoA && orderNoB && orderNoA.length > orderNoB.length) {
      return 1;
    } else {
      if (orderNoA < orderNoB) {
        return -1;
      } else if (orderNoA > orderNoB) {
        return 1;
      } else {
        return 0;
      }
    }
  });

  let mappedMenu: Map<string, any> = new Map();
  sortedPermissionList.forEach((permittedMenu) => {
    if (permittedMenu.accessAll !== '0') {
      let menu = mappedDrawerMenuList.get(permittedMenu.moduleCode);
      if (menu) {
        let drawerMenu = {...permittedMenu, ...menu};
        let [parentModuleCode, childMenuCode] = menu.moduleCode.split('.');

        if (childMenuCode) {
          let childrenMenu = [];
          let parentMenu = mappedMenu.get(parentModuleCode);

          if (parentMenu) {
            let {nested = []} = parentMenu;
            childrenMenu = [...nested];
          }
          childrenMenu.push(drawerMenu);

          mappedMenu.set(parentModuleCode, {
            ...parentMenu,
            nested: childrenMenu,
          });
        } else {
          mappedMenu.set(parentModuleCode, drawerMenu);
        }
      } else {
        mappedMenu.set(permittedMenu.moduleCode, permittedMenu);
      }
    }
  });
  return [...mappedMenu.values()];
}
