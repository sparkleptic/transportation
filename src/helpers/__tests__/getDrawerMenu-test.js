// @flow

import getDrawerMenu from '../getDrawerMenu';

describe('getDrawerMenu', () => {
  it('should return the right menu structure with the right order from the permissions', () => {
    let drawerMenuList = [
      {moduleCode: 'POS', route: '/new-transactions'},
      {
        moduleCode: 'Sales.Overview',
        route: '/sales/overview',
      },
      {
        moduleCode: 'Sales.Transaction',
        route: '/sales/transaction/',
      },
      {
        moduleCode: 'Pickup.Overview',
        route: '/pickup/overview',
      },
      {
        moduleCode: 'Pickup.Request',
        route: '/pickup/request',
      },
      {
        moduleCode: 'Pickup.Schedule',
        route: '/pickup/schedule',
      },
      {
        moduleCode: 'OtherMenu',
        route: '/other/menu',
      },
    ];

    let permissionList = [
      {
        moduleCode: 'POS',
        moduleName: 'Dashboard',
        orderNo: '1',
      },
      {
        moduleCode: 'Pickup',
        moduleName: 'Pickup Item',
        orderNo: '2',
      },
      {
        moduleCode: 'Pickup.Request',
        moduleName: 'Pickup Request',
        orderNo: '2.2',
      },
      {
        moduleCode: 'Pickup.Overview',
        moduleName: 'Pickup Overview',
        orderNo: '2.1',
      },
      {
        moduleCode: 'OtherMenu',
        moduleName: 'Other Menu Item',
        orderNo: '10',
      },
    ];

    let userDrawerMenuList = getDrawerMenu(drawerMenuList, permissionList);

    expect(userDrawerMenuList).toMatchSnapshot();
  });
});
