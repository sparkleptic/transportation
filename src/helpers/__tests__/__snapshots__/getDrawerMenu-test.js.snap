// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getDrawerMenu should return the right menu structure with the right order from the permissions 1`] = `
Array [
  Object {
    "moduleCode": "POS",
    "moduleName": "Dashboard",
    "orderNo": "1",
    "route": "/new-transactions",
  },
  Object {
    "moduleCode": "Pickup",
    "moduleName": "Pickup Item",
    "nested": Array [
      Object {
        "moduleCode": "Pickup.Overview",
        "moduleName": "Pickup Overview",
        "orderNo": "2.1",
        "route": "/pickup/overview",
      },
      Object {
        "moduleCode": "Pickup.Request",
        "moduleName": "Pickup Request",
        "orderNo": "2.2",
        "route": "/pickup/request",
      },
    ],
    "orderNo": "2",
  },
  Object {
    "moduleCode": "OtherMenu",
    "moduleName": "Other Menu Item",
    "orderNo": "10",
    "route": "/other/menu",
  },
]
`;
